File: src/main/java/fundamentals/java/core/Abstraction.java
package fundamentals.java.core;

abstract class Abstraction {
    int i = 1; // mutable state

    abstract int count();

    static void foo() {
        System.out.println("foo");
    }
}

class AbstractionImpl extends Abstraction {
    public int count() {
        foo(); // inherited static method
        return i; // inherited field
    }
    public static void main(String[] args) {
        new AbstractionImpl().count();
    }
}

File: src/main/scala/lang/scala/core/Abstraction.scala
package lang.scala.core

abstract class Abstraction {
  val i = 1 // immutable state

  def count: Int
}

object Abstraction {
  def foo(): Unit = println("foo")
}

class AbstractionImpl extends Abstraction {
  def count: Int = {
    Abstraction.foo() // companion object static method
    i
  }
}

object AbstractionImpl extends App {
  new AbstractionImpl().count
}

